go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,523 @@
+// is the last nad of the network
+func (nadController *netAttachDefinitionController) DeleteNetAttachDef(netAttachDefName string) error {
+	klog.Infof("Delete net-attach-def %s", netAttachDefName)
+	return nadController.perNadNetConfInfo.DoWithLock(netAttachDefName, func(nadName string) error {
+		existingNadNetConfInfo, found := nadController.perNadNetConfInfo.Load(nadName)
+		if !found {
+			klog.V(5).Infof("net-attach-def %s not found", nadName)
+			return nil
+		}
+		err := nadController.deleteNadFromController(existingNadNetConfInfo.netName, nadName)
+		if err != nil {
+			klog.Errorf("Failed to delete net-attach-def %s from network %s: %v", nadName, existingNadNetConfInfo.netName, err)
+			return err
+		}
+		nadController.perNadNetConfInfo.Delete(nadName)
+		return nil
+	})
+}
+
+func (nadController *netAttachDefinitionController) addNadToController(ncm NetworkControllerManager, nadName string,
+	nInfo util.NetInfo, netConfInfo util.NetConfInfo, doStart bool) error {
+	var oc NetworkController
+	var err error
+	var nadExists, isStarted bool
+
+	netName := nInfo.GetNetworkName()
+	klog.V(5).Infof("Add net-attach-def %s to network %s", nadName, netName)
+	return nadController.perNetworkNadNameInfo.DoWithLock(netName, func(networkName string) error {
+		nni, loaded := nadController.perNetworkNadNameInfo.LoadOrStore(networkName, &nadNameInfo{
+			nadNames:  map[string]bool{},
+			oc:        nil,
+			isStarted: false,
+		})
+		if !loaded {
+			// first nad for this network, create controller
+			klog.V(5).Infof("First net-attach-def %s of network %s added, create network controller", nadName, networkName)
+			oc, err = ncm.NewNetworkController(nInfo, netConfInfo)
+			if err != nil {
+				nadController.perNetworkNadNameInfo.Delete(networkName)
+				return fmt.Errorf("failed to create network controller for network %s: %v", networkName, err)
+			}
+			nni.oc = oc
+		} else {
+			klog.V(5).Infof("net-attach-def %s added to existing network %s", nadName, networkName)
+			// controller of this network already exists
+			oc = nni.oc
+			isStarted = nni.isStarted
+			_, nadExists = nni.nadNames[nadName]
+
+			if !oc.CompareNetConf(netConfInfo) {
+				if nadExists {
+					// this should not happen, continue to start the existing controller if requested
+					klog.Warningf("net-attach-def %s netconf spec changed, should not happen", networkName)
+				} else {
+					return fmt.Errorf("nad %s does not share the same CNI config with network %s",
+						nadName, networkName)
+				}
+			}
+		}
+		if !nadExists {
+			nni.nadNames[nadName] = true
+			nni.oc.AddNad(nadName)
+		}
+
+		if !doStart || isStarted {
+			return nil
+		}
+
+		klog.V(5).Infof("Start network controller for network %s", networkName)
+		// start the controller if requested
+		err = oc.Start(context.TODO())
+		if err == nil {
+			nni.isStarted = true
+			return nil
+		}
+
+		if !nadExists {
```

Comment:
can you wrap this in a defer and place it above the start call? Then wrap the oc.Start err with your format and return it. In the defer if err != nil { delete(nni.nadNames, nadName), etc.